modules/stochastic_tools/src/samplers/AdaptiveImportanceSampler.C

```diff
@@ -0,0 +1,196 @@
+    _num_samples_train(getParam<int>("num_samples_train")),
+    _std_factor(getParam<Real>("std_factor")),
+    _use_absolute_value(getParam<bool>("use_absolute_value")),
+    _num_random_seeds(getParam<unsigned int>("num_random_seeds")),
+    _step(getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")->timeStep()),
+    _perf_compute_sample(registerTimedSection("computeSample", 4))
+{
+  // Filling the `distributions` vector with the user-provided distributions.
+  for (const DistributionName & name : _distribution_names)
+    _distributions.push_back(&getDistributionByName(name));
+
+  /* Adaptive Importance Sampling (AdaptiveImportanceSampler) relies on a Markov Chain Monte Carlo
+     (MCMC) algorithm. As such, in MOOSE, any use of MCMC algorithms requires that the `num_steps`
+     parameter in the main App's executioner would control the total number of samples. Therefore,
+     the `num_rows` parameter typically used by exisiting non-MCMC samplers to set the total number
+     of samples has no use here and is fixed to 1.*/
+  setNumberOfRows(1);
+
+  // Setting the number of columns in the sampler matrix (equal to the number of distributions).
+  setNumberOfCols(_distributions.size());
+
+  /* `inputs_sto` is a member variable that aids in forming the importance distribution.
+     One dimension of this variable is equal to the number of distributions. The other dimension
+     of the variable, at the last step, is equal to the number of samples the user desires.*/
+  _inputs_sto.resize(_distributions.size());
+
+  // Mapping all the input distributions to a standard normal space
+  for (unsigned int i = 0; i < _distributions.size(); ++i)
+    _inputs_sto[i].push_back(Normal::quantile(_distributions[i]->cdf(_initial_values[i]), 0, 1));
+
+  /* `prev_value` is a member variable for tracking the previously accepted samples in the
+     MCMC algorithm and proposing the next sample.*/
+  _prev_value.resize(_distributions.size());
+
+  // `check_step` is a member variable for ensuring that the MCMC algorithm proceeds in a sequential
+  // fashion.
+  _check_step = 0;
+
+  // Storage for means of input values for proposing the next sample
+  _mean_sto.resize(_distributions.size());
+
+  // Storage for standard deviations of input values for proposing the next sample
+  _std_sto.resize(_distributions.size());
+
+  setNumberOfRandomSeeds(_num_random_seeds);
+}
+
+const std::vector<Real> &
+AdaptiveImportanceSampler::getInitialValues() const
+{
+  return _initial_values;
+}
+
+const int &
+AdaptiveImportanceSampler::getNumSamplesTrain() const
+{
+  return _num_samples_train;
+}
+
+const bool &
+AdaptiveImportanceSampler::getUseAbsoluteValue() const
+{
+  return _use_absolute_value;
+}
+
+const Real &
+AdaptiveImportanceSampler::getOutputLimit() const
+{
+  return _output_limit;
+}
+
+Real
+AdaptiveImportanceSampler::computeSample(dof_id_type /*row_index*/, dof_id_type col_index)
+{
+  TIME_SECTION(_perf_compute_sample);
```

Comment:
I would remove this since this happens way too many times and it will cause a slow down.